File: src/main/java/com/kevinmost/command/CommandExecutor.java
package com.kevinmost.command;

import com.kevinmost.PlatformDetails;
import org.apache.commons.io.Charsets;
import org.apache.commons.io.IOUtils;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;

public class CommandExecutor {
    private static final Runtime RUNTIME = Runtime.getRuntime();

    public static <T> T executeCommand(TerminalCommand<T> command, String... args) {
        try {
            final String resourceFolderPath = PlatformDetails.get().osResourceFolder.getPath() + File.separatorChar + command.getTerminalCommand(args);
            final Process result = RUNTIME.exec(resourceFolderPath);
            final String stdOut = streamToString(result.getInputStream());
            final String stdErr = streamToString(result.getErrorStream());
            return command.transformToResult(stdOut, stdErr);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    private static String streamToString(InputStream stream) {
        try {
            return IOUtils.toString(stream, Charsets.UTF_8);
        } catch (IOException e) {
            e.printStackTrace();
            return "";
        } finally {
            IOUtils.closeQuietly(stream);
        }
    }
}

File: src/main/java/com/kevinmost/command/CommandExecutor.kt
package com.kevinmost.command

import java.io.InputStreamReader

public fun <RESULT> executeCommand(
        command: TerminalCommand, transformer: CommandOutputTransformer<RESULT>, vararg args: String) : RESULT {
    val runtime = Runtime.getRuntime();
    // TODO: Why don't the varargs go into getTerminalCommand properly?
    val exec = runtime.exec(command.getTerminalCommand(args));
    val stdout : String = InputStreamReader(exec.getInputStream()).readText();
    val stderr : String = InputStreamReader(exec.getErrorStream()).readText();
    return transformer.getResult(stdout, stderr);
}
